### PR TITLE
Wallet thread safe with locks

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

File: Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs [hunk 1]
```diff
@@ -302,85 +302,85 @@ public IActionResult GetHistory([FromQuery] WalletHistoryRequest request)
...
-                // in order to calculate the fee properly we need to retrieve all the transactions with spending details.
-                var wallet = this.walletManager.GetWalletByName(request.WalletName);
-                var allTransactionsWithSpendingDetails = wallet.GetAllTransactionsByCoinType(this.coinType).Where(t => t.SpendingDetails != null).ToList();
-
                 // get transactions contained in the wallet
-                var addresses = this.walletManager.GetHistory(request.WalletName).ToList();
-                foreach (var address in addresses)
+                var items = this.walletManager.GetFlatHistory(request.WalletName).ToList().OrderByDescending(o => o.Transaction.CreationTime).Take(100).ToList();
+                List<FlatHistory> spendingDetails = items.Where(t => t.Transaction.SpendingDetails != null).ToList();
+                List<FlatHistory> filttered = items.Where(t => !t.Address.IsChangeAddress() || (t.Address.IsChangeAddress() && !t.Transaction.IsSpendable())).ToList();
```
Comment (mikedennis) [inline Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs]: filttered => filtered

File: Stratis.Bitcoin.Features.Wallet/Interfaces/IWalletManager.cs [hunk 2]
```diff
@@ -112,6 +112,13 @@ public interface IWalletManager
         IEnumerable<HdAddress> GetHistory(string walletName);
 
         /// <summary>
+        /// Return a flat representation of the wallet history.
+        /// </summary>
+        /// <param name="walletName">The wallet name.</param>
+        /// <returns>The history in a address to transaction pairs.</returns>
```
Comment (mikedennis) [inline Stratis.Bitcoin.Features.Wallet/Interfaces/IWalletManager.cs]: Not sure a better alternative by I find this comment confusing. Maybe this "Collection of address history and transaction pairs."?

Comment (dangershony) [reply Stratis.Bitcoin.Features.Wallet/Interfaces/IWalletManager.cs]: Ha yeah totally confusing now that I read it yap I will fix

File: Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs (hunk 1 above)
Comment (dangershony) [reply Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs]: Will fix thanks

**Comment by bokobza** [comment]
I updated my comments here
https://github.com/stratisproject/StratisBitcoinFullNode/commit/61c3b92f12afd3fc2b3904af990f73781a6ede69

**Pull request merged**